{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "import requests\n",
    "\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_histogram_per_predclass(df, df_name):\n",
    "    missing_percentage_per_class = {}\n",
    "    total_rows_per_class = df.groupby('PredClass').size()\n",
    "    \n",
    "    for pred_class, group in df.groupby('PredClass'):\n",
    "        missing_rows = group['Pred'].isnull().sum()\n",
    "        missing_percentage = (missing_rows / total_rows_per_class[pred_class]) * 100\n",
    "        missing_percentage_per_class[pred_class] = missing_percentage\n",
    "\n",
    "    print(f\"\\n{df_name} - Percentage of missing 'Pred' values per 'PredClass':\")\n",
    "    for pred_class, missing_percentage in missing_percentage_per_class.items():\n",
    "        print(f\"{pred_class}: {missing_percentage:.2f}%\")\n",
    "\n",
    "    print(f\"\\n{df_name} - Number of rows per 'PredClass' where 'Pred' value is present:\")\n",
    "    for pred_class, group in df.groupby('PredClass'):\n",
    "        rows_with_pred = group['Pred'].notnull().sum()\n",
    "        print(f\"{pred_class}: {rows_with_pred}\")\n",
    "\n",
    "        sorted_pred = group['Pred'].dropna().sort_values()   \n",
    "        print(f\"Top 5 highest 'Pred' values for {pred_class}: {sorted_pred[-5:].values}\")\n",
    "        print(f\"Bottom 5 lowest 'Pred' values for {pred_class}: {sorted_pred[:5].values}\")\n",
    "\n",
    "    df_dropna = df.dropna(subset=['Pred'])\n",
    "    plt.figure(figsize=(12, 6))\n",
    "    for pred_class, group in df_dropna.groupby('PredClass'):\n",
    "        plt.hist(group['Pred'], bins=20, alpha=0.5, label=str(pred_class))\n",
    "\n",
    "    plt.xlabel('Pred Value')\n",
    "    plt.ylabel('Frequency')\n",
    "    plt.title(f'Histogram of Pred Values per PredClass for {df_name}')\n",
    "    plt.legend(title='Pred Class', loc='upper right')\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Imports "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = '/Users/kristinagrigaityte/PycharmProjects/pulls/Files/BSModel/Predictions'\n",
    "csv_files = [file for file in os.listdir(directory) if file.endswith('.csv')]\n",
    "\n",
    "for file in csv_files:\n",
    "    df_name = os.path.splitext(file)[0] \n",
    "    globals()[df_name] = pd.read_csv(os.path.join(directory, file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>recent_selection</th>\n",
       "      <th>recent_selection_avg</th>\n",
       "      <th>recent_balancingSelection</th>\n",
       "      <th>recent_balancingSelection_avg</th>\n",
       "      <th>recent_negativeFreq-dep.Selection</th>\n",
       "      <th>recent_negativeFreq-dep.Selection_avg</th>\n",
       "      <th>recent_overdominance</th>\n",
       "      <th>old_selection</th>\n",
       "      <th>old_selection_avg</th>\n",
       "      <th>old_balancingSelection</th>\n",
       "      <th>old_balancingSelection_avg</th>\n",
       "      <th>old_negativeFreq-dep.Selection</th>\n",
       "      <th>old_negativeFreq-dep.Selection_avg</th>\n",
       "      <th>old_overdominance</th>\n",
       "      <th>medium_selection</th>\n",
       "      <th>medium_selection_avg</th>\n",
       "      <th>medium_balancingSelection</th>\n",
       "      <th>medium_balancingSelection_avg</th>\n",
       "      <th>medium_negativeFreq-dep.Selection</th>\n",
       "      <th>medium_negativeFreq-dep.Selection_avg</th>\n",
       "      <th>medium_overdominance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ch1</td>\n",
       "      <td>150</td>\n",
       "      <td>0.998</td>\n",
       "      <td>23</td>\n",
       "      <td>0.606</td>\n",
       "      <td>147</td>\n",
       "      <td>0.565</td>\n",
       "      <td>8583</td>\n",
       "      <td>150</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>150</td>\n",
       "      <td>0.655</td>\n",
       "      <td>8580</td>\n",
       "      <td>150</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>150</td>\n",
       "      <td>0.709</td>\n",
       "      <td>8580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ch2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.985</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.526</td>\n",
       "      <td>9060</td>\n",
       "      <td>1</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.549</td>\n",
       "      <td>9060</td>\n",
       "      <td>1</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.575</td>\n",
       "      <td>9060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ch3</td>\n",
       "      <td>763</td>\n",
       "      <td>0.998</td>\n",
       "      <td>106</td>\n",
       "      <td>0.619</td>\n",
       "      <td>738</td>\n",
       "      <td>0.552</td>\n",
       "      <td>6633</td>\n",
       "      <td>763</td>\n",
       "      <td>0.998</td>\n",
       "      <td>10</td>\n",
       "      <td>0.602</td>\n",
       "      <td>763</td>\n",
       "      <td>0.628</td>\n",
       "      <td>6608</td>\n",
       "      <td>763</td>\n",
       "      <td>1.000</td>\n",
       "      <td>17</td>\n",
       "      <td>0.600</td>\n",
       "      <td>763</td>\n",
       "      <td>0.678</td>\n",
       "      <td>6608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ch4</td>\n",
       "      <td>454</td>\n",
       "      <td>0.999</td>\n",
       "      <td>58</td>\n",
       "      <td>0.594</td>\n",
       "      <td>453</td>\n",
       "      <td>0.573</td>\n",
       "      <td>5730</td>\n",
       "      <td>454</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>454</td>\n",
       "      <td>0.676</td>\n",
       "      <td>5729</td>\n",
       "      <td>454</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>454</td>\n",
       "      <td>0.736</td>\n",
       "      <td>5729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ch5</td>\n",
       "      <td>133</td>\n",
       "      <td>0.999</td>\n",
       "      <td>10</td>\n",
       "      <td>0.579</td>\n",
       "      <td>133</td>\n",
       "      <td>0.573</td>\n",
       "      <td>6339</td>\n",
       "      <td>133</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>133</td>\n",
       "      <td>0.672</td>\n",
       "      <td>6339</td>\n",
       "      <td>133</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>133</td>\n",
       "      <td>0.732</td>\n",
       "      <td>6339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ch6</td>\n",
       "      <td>1393</td>\n",
       "      <td>0.998</td>\n",
       "      <td>252</td>\n",
       "      <td>0.601</td>\n",
       "      <td>1375</td>\n",
       "      <td>0.562</td>\n",
       "      <td>5693</td>\n",
       "      <td>1393</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1</td>\n",
       "      <td>0.520</td>\n",
       "      <td>1393</td>\n",
       "      <td>0.652</td>\n",
       "      <td>5675</td>\n",
       "      <td>1393</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.509</td>\n",
       "      <td>1393</td>\n",
       "      <td>0.704</td>\n",
       "      <td>5675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ch7</td>\n",
       "      <td>353</td>\n",
       "      <td>0.999</td>\n",
       "      <td>28</td>\n",
       "      <td>0.599</td>\n",
       "      <td>350</td>\n",
       "      <td>0.580</td>\n",
       "      <td>5583</td>\n",
       "      <td>353</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>353</td>\n",
       "      <td>0.687</td>\n",
       "      <td>5580</td>\n",
       "      <td>353</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>353</td>\n",
       "      <td>0.752</td>\n",
       "      <td>5580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ch8</td>\n",
       "      <td>145</td>\n",
       "      <td>0.994</td>\n",
       "      <td>20</td>\n",
       "      <td>0.567</td>\n",
       "      <td>144</td>\n",
       "      <td>0.546</td>\n",
       "      <td>5590</td>\n",
       "      <td>145</td>\n",
       "      <td>0.992</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>145</td>\n",
       "      <td>0.614</td>\n",
       "      <td>5589</td>\n",
       "      <td>145</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>145</td>\n",
       "      <td>0.662</td>\n",
       "      <td>5589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ch9</td>\n",
       "      <td>390</td>\n",
       "      <td>0.999</td>\n",
       "      <td>54</td>\n",
       "      <td>0.629</td>\n",
       "      <td>385</td>\n",
       "      <td>0.579</td>\n",
       "      <td>4579</td>\n",
       "      <td>390</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1</td>\n",
       "      <td>0.509</td>\n",
       "      <td>390</td>\n",
       "      <td>0.684</td>\n",
       "      <td>4574</td>\n",
       "      <td>390</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>390</td>\n",
       "      <td>0.744</td>\n",
       "      <td>4574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ch10</td>\n",
       "      <td>429</td>\n",
       "      <td>0.999</td>\n",
       "      <td>74</td>\n",
       "      <td>0.615</td>\n",
       "      <td>425</td>\n",
       "      <td>0.584</td>\n",
       "      <td>5455</td>\n",
       "      <td>429</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>429</td>\n",
       "      <td>0.700</td>\n",
       "      <td>5451</td>\n",
       "      <td>429</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>429</td>\n",
       "      <td>0.759</td>\n",
       "      <td>5451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ch11</td>\n",
       "      <td>375</td>\n",
       "      <td>0.999</td>\n",
       "      <td>52</td>\n",
       "      <td>0.586</td>\n",
       "      <td>373</td>\n",
       "      <td>0.578</td>\n",
       "      <td>5323</td>\n",
       "      <td>375</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>375</td>\n",
       "      <td>0.686</td>\n",
       "      <td>5321</td>\n",
       "      <td>375</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>375</td>\n",
       "      <td>0.745</td>\n",
       "      <td>5321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ch12</td>\n",
       "      <td>555</td>\n",
       "      <td>0.999</td>\n",
       "      <td>48</td>\n",
       "      <td>0.599</td>\n",
       "      <td>550</td>\n",
       "      <td>0.588</td>\n",
       "      <td>4849</td>\n",
       "      <td>555</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>555</td>\n",
       "      <td>0.699</td>\n",
       "      <td>4844</td>\n",
       "      <td>555</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.506</td>\n",
       "      <td>555</td>\n",
       "      <td>0.761</td>\n",
       "      <td>4844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ch13</td>\n",
       "      <td>1635</td>\n",
       "      <td>0.998</td>\n",
       "      <td>397</td>\n",
       "      <td>0.619</td>\n",
       "      <td>1604</td>\n",
       "      <td>0.556</td>\n",
       "      <td>2525</td>\n",
       "      <td>1635</td>\n",
       "      <td>0.998</td>\n",
       "      <td>11</td>\n",
       "      <td>0.552</td>\n",
       "      <td>1635</td>\n",
       "      <td>0.641</td>\n",
       "      <td>2494</td>\n",
       "      <td>1635</td>\n",
       "      <td>1.000</td>\n",
       "      <td>24</td>\n",
       "      <td>0.672</td>\n",
       "      <td>1635</td>\n",
       "      <td>0.689</td>\n",
       "      <td>2494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ch14</td>\n",
       "      <td>7</td>\n",
       "      <td>0.956</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>7</td>\n",
       "      <td>0.525</td>\n",
       "      <td>3500</td>\n",
       "      <td>7</td>\n",
       "      <td>0.944</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>7</td>\n",
       "      <td>0.548</td>\n",
       "      <td>3500</td>\n",
       "      <td>7</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>7</td>\n",
       "      <td>0.584</td>\n",
       "      <td>3500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ch15</td>\n",
       "      <td>1059</td>\n",
       "      <td>0.999</td>\n",
       "      <td>136</td>\n",
       "      <td>0.609</td>\n",
       "      <td>1048</td>\n",
       "      <td>0.578</td>\n",
       "      <td>2389</td>\n",
       "      <td>1059</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1059</td>\n",
       "      <td>0.680</td>\n",
       "      <td>2378</td>\n",
       "      <td>1059</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1059</td>\n",
       "      <td>0.737</td>\n",
       "      <td>2378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ch16</td>\n",
       "      <td>913</td>\n",
       "      <td>1.000</td>\n",
       "      <td>116</td>\n",
       "      <td>0.593</td>\n",
       "      <td>913</td>\n",
       "      <td>0.606</td>\n",
       "      <td>2625</td>\n",
       "      <td>913</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.516</td>\n",
       "      <td>913</td>\n",
       "      <td>0.741</td>\n",
       "      <td>2625</td>\n",
       "      <td>913</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>913</td>\n",
       "      <td>0.802</td>\n",
       "      <td>2625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ch17</td>\n",
       "      <td>531</td>\n",
       "      <td>0.998</td>\n",
       "      <td>62</td>\n",
       "      <td>0.589</td>\n",
       "      <td>529</td>\n",
       "      <td>0.581</td>\n",
       "      <td>2573</td>\n",
       "      <td>531</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>531</td>\n",
       "      <td>0.684</td>\n",
       "      <td>2571</td>\n",
       "      <td>531</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>531</td>\n",
       "      <td>0.745</td>\n",
       "      <td>2571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ch18</td>\n",
       "      <td>714</td>\n",
       "      <td>0.999</td>\n",
       "      <td>111</td>\n",
       "      <td>0.588</td>\n",
       "      <td>695</td>\n",
       "      <td>0.575</td>\n",
       "      <td>2572</td>\n",
       "      <td>714</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>714</td>\n",
       "      <td>0.674</td>\n",
       "      <td>2553</td>\n",
       "      <td>714</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>714</td>\n",
       "      <td>0.730</td>\n",
       "      <td>2553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ch19</td>\n",
       "      <td>951</td>\n",
       "      <td>0.998</td>\n",
       "      <td>140</td>\n",
       "      <td>0.577</td>\n",
       "      <td>933</td>\n",
       "      <td>0.560</td>\n",
       "      <td>1315</td>\n",
       "      <td>951</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>951</td>\n",
       "      <td>0.645</td>\n",
       "      <td>1297</td>\n",
       "      <td>951</td>\n",
       "      <td>1.000</td>\n",
       "      <td>5</td>\n",
       "      <td>0.506</td>\n",
       "      <td>951</td>\n",
       "      <td>0.700</td>\n",
       "      <td>1297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ch20</td>\n",
       "      <td>458</td>\n",
       "      <td>0.999</td>\n",
       "      <td>58</td>\n",
       "      <td>0.596</td>\n",
       "      <td>452</td>\n",
       "      <td>0.584</td>\n",
       "      <td>2376</td>\n",
       "      <td>458</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1</td>\n",
       "      <td>0.511</td>\n",
       "      <td>458</td>\n",
       "      <td>0.695</td>\n",
       "      <td>2370</td>\n",
       "      <td>458</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>458</td>\n",
       "      <td>0.756</td>\n",
       "      <td>2370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ch21</td>\n",
       "      <td>738</td>\n",
       "      <td>0.999</td>\n",
       "      <td>109</td>\n",
       "      <td>0.597</td>\n",
       "      <td>729</td>\n",
       "      <td>0.573</td>\n",
       "      <td>789</td>\n",
       "      <td>738</td>\n",
       "      <td>0.999</td>\n",
       "      <td>1</td>\n",
       "      <td>0.507</td>\n",
       "      <td>738</td>\n",
       "      <td>0.672</td>\n",
       "      <td>780</td>\n",
       "      <td>738</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.511</td>\n",
       "      <td>738</td>\n",
       "      <td>0.729</td>\n",
       "      <td>780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ch22</td>\n",
       "      <td>1224</td>\n",
       "      <td>0.999</td>\n",
       "      <td>177</td>\n",
       "      <td>0.610</td>\n",
       "      <td>1206</td>\n",
       "      <td>0.581</td>\n",
       "      <td>357</td>\n",
       "      <td>1224</td>\n",
       "      <td>0.999</td>\n",
       "      <td>3</td>\n",
       "      <td>0.536</td>\n",
       "      <td>1224</td>\n",
       "      <td>0.688</td>\n",
       "      <td>339</td>\n",
       "      <td>1224</td>\n",
       "      <td>1.000</td>\n",
       "      <td>6</td>\n",
       "      <td>0.531</td>\n",
       "      <td>1224</td>\n",
       "      <td>0.747</td>\n",
       "      <td>339</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  recent_selection recent_selection_avg  \\\n",
       "0         ch1               150                0.998   \n",
       "1         ch2                 1                0.985   \n",
       "2         ch3               763                0.998   \n",
       "3         ch4               454                0.999   \n",
       "4         ch5               133                0.999   \n",
       "5         ch6              1393                0.998   \n",
       "6         ch7               353                0.999   \n",
       "7         ch8               145                0.994   \n",
       "8         ch9               390                0.999   \n",
       "9        ch10               429                0.999   \n",
       "10       ch11               375                0.999   \n",
       "11       ch12               555                0.999   \n",
       "12       ch13              1635                0.998   \n",
       "13       ch14                 7                0.956   \n",
       "14       ch15              1059                0.999   \n",
       "15       ch16               913                1.000   \n",
       "16       ch17               531                0.998   \n",
       "17       ch18               714                0.999   \n",
       "18       ch19               951                0.998   \n",
       "19       ch20               458                0.999   \n",
       "20       ch21               738                0.999   \n",
       "21       ch22              1224                0.999   \n",
       "\n",
       "    recent_balancingSelection recent_balancingSelection_avg  \\\n",
       "0                          23                         0.606   \n",
       "1                           0                         0.000   \n",
       "2                         106                         0.619   \n",
       "3                          58                         0.594   \n",
       "4                          10                         0.579   \n",
       "5                         252                         0.601   \n",
       "6                          28                         0.599   \n",
       "7                          20                         0.567   \n",
       "8                          54                         0.629   \n",
       "9                          74                         0.615   \n",
       "10                         52                         0.586   \n",
       "11                         48                         0.599   \n",
       "12                        397                         0.619   \n",
       "13                          0                         0.000   \n",
       "14                        136                         0.609   \n",
       "15                        116                         0.593   \n",
       "16                         62                         0.589   \n",
       "17                        111                         0.588   \n",
       "18                        140                         0.577   \n",
       "19                         58                         0.596   \n",
       "20                        109                         0.597   \n",
       "21                        177                         0.610   \n",
       "\n",
       "    recent_negativeFreq-dep.Selection recent_negativeFreq-dep.Selection_avg  \\\n",
       "0                                 147                                 0.565   \n",
       "1                                   1                                 0.526   \n",
       "2                                 738                                 0.552   \n",
       "3                                 453                                 0.573   \n",
       "4                                 133                                 0.573   \n",
       "5                                1375                                 0.562   \n",
       "6                                 350                                 0.580   \n",
       "7                                 144                                 0.546   \n",
       "8                                 385                                 0.579   \n",
       "9                                 425                                 0.584   \n",
       "10                                373                                 0.578   \n",
       "11                                550                                 0.588   \n",
       "12                               1604                                 0.556   \n",
       "13                                  7                                 0.525   \n",
       "14                               1048                                 0.578   \n",
       "15                                913                                 0.606   \n",
       "16                                529                                 0.581   \n",
       "17                                695                                 0.575   \n",
       "18                                933                                 0.560   \n",
       "19                                452                                 0.584   \n",
       "20                                729                                 0.573   \n",
       "21                               1206                                 0.581   \n",
       "\n",
       "    recent_overdominance  old_selection old_selection_avg  \\\n",
       "0                   8583            150             0.997   \n",
       "1                   9060              1             0.986   \n",
       "2                   6633            763             0.998   \n",
       "3                   5730            454             0.999   \n",
       "4                   6339            133             0.999   \n",
       "5                   5693           1393             0.999   \n",
       "6                   5583            353             0.999   \n",
       "7                   5590            145             0.992   \n",
       "8                   4579            390             0.999   \n",
       "9                   5455            429             0.999   \n",
       "10                  5323            375             0.999   \n",
       "11                  4849            555             0.999   \n",
       "12                  2525           1635             0.998   \n",
       "13                  3500              7             0.944   \n",
       "14                  2389           1059             0.999   \n",
       "15                  2625            913             1.000   \n",
       "16                  2573            531             0.999   \n",
       "17                  2572            714             0.999   \n",
       "18                  1315            951             0.998   \n",
       "19                  2376            458             0.999   \n",
       "20                   789            738             0.999   \n",
       "21                   357           1224             0.999   \n",
       "\n",
       "    old_balancingSelection old_balancingSelection_avg  \\\n",
       "0                        0                      0.000   \n",
       "1                        0                      0.000   \n",
       "2                       10                      0.602   \n",
       "3                        0                      0.000   \n",
       "4                        0                      0.000   \n",
       "5                        1                      0.520   \n",
       "6                        0                      0.000   \n",
       "7                        0                      0.000   \n",
       "8                        1                      0.509   \n",
       "9                        0                      0.000   \n",
       "10                       0                      0.000   \n",
       "11                       0                      0.000   \n",
       "12                      11                      0.552   \n",
       "13                       0                      0.000   \n",
       "14                       0                      0.000   \n",
       "15                       1                      0.516   \n",
       "16                       0                      0.000   \n",
       "17                       0                      0.000   \n",
       "18                       0                      0.000   \n",
       "19                       1                      0.511   \n",
       "20                       1                      0.507   \n",
       "21                       3                      0.536   \n",
       "\n",
       "    old_negativeFreq-dep.Selection old_negativeFreq-dep.Selection_avg  \\\n",
       "0                              150                              0.655   \n",
       "1                                1                              0.549   \n",
       "2                              763                              0.628   \n",
       "3                              454                              0.676   \n",
       "4                              133                              0.672   \n",
       "5                             1393                              0.652   \n",
       "6                              353                              0.687   \n",
       "7                              145                              0.614   \n",
       "8                              390                              0.684   \n",
       "9                              429                              0.700   \n",
       "10                             375                              0.686   \n",
       "11                             555                              0.699   \n",
       "12                            1635                              0.641   \n",
       "13                               7                              0.548   \n",
       "14                            1059                              0.680   \n",
       "15                             913                              0.741   \n",
       "16                             531                              0.684   \n",
       "17                             714                              0.674   \n",
       "18                             951                              0.645   \n",
       "19                             458                              0.695   \n",
       "20                             738                              0.672   \n",
       "21                            1224                              0.688   \n",
       "\n",
       "    old_overdominance  medium_selection medium_selection_avg  \\\n",
       "0                8580               150                1.000   \n",
       "1                9060                 1                0.998   \n",
       "2                6608               763                1.000   \n",
       "3                5729               454                1.000   \n",
       "4                6339               133                1.000   \n",
       "5                5675              1393                1.000   \n",
       "6                5580               353                1.000   \n",
       "7                5589               145                0.999   \n",
       "8                4574               390                1.000   \n",
       "9                5451               429                1.000   \n",
       "10               5321               375                1.000   \n",
       "11               4844               555                1.000   \n",
       "12               2494              1635                1.000   \n",
       "13               3500                 7                0.990   \n",
       "14               2378              1059                1.000   \n",
       "15               2625               913                1.000   \n",
       "16               2571               531                1.000   \n",
       "17               2553               714                1.000   \n",
       "18               1297               951                1.000   \n",
       "19               2370               458                1.000   \n",
       "20                780               738                1.000   \n",
       "21                339              1224                1.000   \n",
       "\n",
       "    medium_balancingSelection medium_balancingSelection_avg  \\\n",
       "0                           0                         0.000   \n",
       "1                           0                         0.000   \n",
       "2                          17                         0.600   \n",
       "3                           0                         0.000   \n",
       "4                           0                         0.000   \n",
       "5                           1                         0.509   \n",
       "6                           0                         0.000   \n",
       "7                           0                         0.000   \n",
       "8                           0                         0.000   \n",
       "9                           0                         0.000   \n",
       "10                          0                         0.000   \n",
       "11                          1                         0.506   \n",
       "12                         24                         0.672   \n",
       "13                          0                         0.000   \n",
       "14                          0                         0.000   \n",
       "15                          0                         0.000   \n",
       "16                          0                         0.000   \n",
       "17                          0                         0.000   \n",
       "18                          5                         0.506   \n",
       "19                          0                         0.000   \n",
       "20                          1                         0.511   \n",
       "21                          6                         0.531   \n",
       "\n",
       "    medium_negativeFreq-dep.Selection medium_negativeFreq-dep.Selection_avg  \\\n",
       "0                                 150                                 0.709   \n",
       "1                                   1                                 0.575   \n",
       "2                                 763                                 0.678   \n",
       "3                                 454                                 0.736   \n",
       "4                                 133                                 0.732   \n",
       "5                                1393                                 0.704   \n",
       "6                                 353                                 0.752   \n",
       "7                                 145                                 0.662   \n",
       "8                                 390                                 0.744   \n",
       "9                                 429                                 0.759   \n",
       "10                                375                                 0.745   \n",
       "11                                555                                 0.761   \n",
       "12                               1635                                 0.689   \n",
       "13                                  7                                 0.584   \n",
       "14                               1059                                 0.737   \n",
       "15                                913                                 0.802   \n",
       "16                                531                                 0.745   \n",
       "17                                714                                 0.730   \n",
       "18                                951                                 0.700   \n",
       "19                                458                                 0.756   \n",
       "20                                738                                 0.729   \n",
       "21                               1224                                 0.747   \n",
       "\n",
       "    medium_overdominance  \n",
       "0                   8580  \n",
       "1                   9060  \n",
       "2                   6608  \n",
       "3                   5729  \n",
       "4                   6339  \n",
       "5                   5675  \n",
       "6                   5580  \n",
       "7                   5589  \n",
       "8                   4574  \n",
       "9                   5451  \n",
       "10                  5321  \n",
       "11                  4844  \n",
       "12                  2494  \n",
       "13                  3500  \n",
       "14                  2378  \n",
       "15                  2625  \n",
       "16                  2571  \n",
       "17                  2553  \n",
       "18                  1297  \n",
       "19                  2370  \n",
       "20                   780  \n",
       "21                   339  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "categories = ['Selection', 'Balancing Selection', 'Negative Freq-Dep. Selection', 'Overdominance']\n",
    "prefixes = ['recent', 'old', 'medium']\n",
    "indexes = ['1', '2', '3', '3']\n",
    "chromosomes = ['ch1', 'ch2', 'ch3', 'ch4', 'ch5', 'ch6', 'ch7', 'ch8', 'ch9', 'ch10',\n",
    "               'ch11', 'ch12', 'ch13', 'ch14', 'ch15', 'ch16', 'ch17', 'ch18', 'ch19', 'ch20', 'ch21', 'ch22']\n",
    "\n",
    "data_list = []\n",
    "\n",
    "for chrom in chromosomes:\n",
    "    count_dict = {'chromosome': chrom}\n",
    "    \n",
    "    for prefix in prefixes:\n",
    "        for i, category in zip(indexes, categories):\n",
    "            category_name_parts = category.split()\n",
    "            formatted_category_name = category_name_parts[0].lower() + ''.join(word.capitalize() for word in category_name_parts[1:])\n",
    "            df_name = f'prediction_{prefix}_{i}_{chrom}'\n",
    "            df = globals()[df_name]\n",
    "            \n",
    "            selected_rows = df[df['PredClass'] == category]\n",
    "            \n",
    "            count = selected_rows.shape[0]\n",
    "            count_dict[f'{prefix}_{formatted_category_name}'] = count\n",
    "            \n",
    "            if category != 'Overdominance':\n",
    "                if count > 0:\n",
    "                    avg_score = selected_rows['Pred'].mean()\n",
    "                else:\n",
    "                    avg_score = 0.0\n",
    "                count_dict[f'{prefix}_{formatted_category_name}_avg'] = f\"{avg_score:.3f}\"\n",
    "    \n",
    "    data_list.append(count_dict)\n",
    "\n",
    "counts_df = pd.DataFrame(data_list)\n",
    "pd.set_option('display.max_columns', None)\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def plot_line_with_values(dataframe, prefix, legend_labels, colors, filename):\n",
    "    relevant_columns = [col for col in dataframe.columns if col.startswith(prefix) and \"_avg\" in col]\n",
    "    plot_data = dataframe[['chromosome'] + relevant_columns]\n",
    "    \n",
    "    plot_data.set_index('chromosome', inplace=True)\n",
    "    plot_data = plot_data.apply(pd.to_numeric, errors='coerce')\n",
    "    plot_data.dropna(subset=relevant_columns, how='all', inplace=True)\n",
    "    \n",
    "    ax = plot_data.plot(kind='line', figsize=(14, 7), color=colors, marker='o')\n",
    "    plt.title(f'Line Plot for {prefix.capitalize()} Selection')\n",
    "    plt.xlabel('Chromosome')\n",
    "    plt.ylabel('Prediction Score')\n",
    "    \n",
    "    new_labels = [legend_labels.get(col, col) for col in relevant_columns]\n",
    "    plt.legend(new_labels, title='Selection Type', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "    \n",
    "    ax.set_xticks(range(len(plot_data.index)))\n",
    "    ax.set_xticklabels(plot_data.index, rotation=45, ha='right')\n",
    "    \n",
    "    for line in ax.get_lines():\n",
    "        for x, y in zip(line.get_xdata(), line.get_ydata()):\n",
    "            if y == 0:\n",
    "                ax.annotate(f'{y:.2f}', xy=(x, y), xytext=(5, 5), textcoords='offset points')\n",
    "\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    plt.savefig(filename)\n",
    "    plt.close()\n",
    "\n",
    "legend_labels_avg = {\n",
    "    'recent_selection_avg': 'Selection Prediction Average',\n",
    "    'recent_balancingSelection_avg': 'Balancing Selection Prediction Average',\n",
    "    'recent_negativeFreq-dep.Selection_avg': 'Negative Freq-dep. Selection Prediction Average',\n",
    "    \n",
    "    'old_selection_avg': 'Selection Prediction Average',\n",
    "    'old_balancingSelection_avg': 'Balancing Selection Prediction Average',\n",
    "    'old_negativeFreq-dep.Selection_avg': 'Negative Freq-dep. Selection Prediction Average',\n",
    "    \n",
    "    'medium_selection_avg': 'Selection Prediction Average',\n",
    "    'medium_balancingSelection_avg': 'Balancing Selection Prediction Average',\n",
    "    'medium_negativeFreq-dep.Selection_avg': 'Negative Freq-dep. Selection Prediction Average'\n",
    "}\n",
    "\n",
    "secure_blue = '#3357A4'\n",
    "peaceful_mint = '#A0E0D8'\n",
    "innovation_purple = '#879BEF'\n",
    "\n",
    "colors_dict = {\n",
    "    'old': [secure_blue, peaceful_mint, innovation_purple],\n",
    "    'recent': [secure_blue, peaceful_mint, innovation_purple],\n",
    "    'medium': [secure_blue, peaceful_mint, innovation_purple]\n",
    "}\n",
    "\n",
    "for prefix in ['old', 'recent', 'medium']:\n",
    "    filename = f'{prefix}_selection_plot.png'\n",
    "    plot_line_with_values(counts_df, prefix, legend_labels_avg, colors_dict[prefix], filename)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_stacked_bars(dataframe, prefix, legend_labels, colors, filename):\n",
    "    relevant_columns = [col for col in dataframe.columns if col.startswith(prefix) and \"avg\" not in col]\n",
    "    plot_data = dataframe[['chromosome'] + relevant_columns]\n",
    "\n",
    "    plot_data.set_index('chromosome', inplace=True)\n",
    "\n",
    "    ax = plot_data.plot(kind='bar', stacked=True, figsize=(14, 7), color=colors)\n",
    "    plt.title(f'Stacked Bar Plot for {prefix.capitalize()} Selection')\n",
    "    plt.xlabel('Chromosome')\n",
    "    plt.ylabel('SNP Count')\n",
    "    \n",
    "    new_labels = [legend_labels.get(col, col) for col in relevant_columns]\n",
    "    \n",
    "    plt.legend(new_labels, title='Selection Type', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    plt.savefig(filename)\n",
    "    plt.close()\n",
    "\n",
    "legend_labels = {\n",
    "    'recent_selection': 'Selection',\n",
    "    'recent_balancingSelection': 'Balancing Selection',\n",
    "    'recent_negativeFreq-dep.Selection': 'Negative Freq-dep. Selection',\n",
    "    'recent_overdominance': 'Overdominance',\n",
    "    \n",
    "    'old_selection': 'Selection',\n",
    "    'old_balancingSelection': 'Balancing Selection',\n",
    "    'old_negativeFreq-dep.Selection': 'Negative Freq-dep. Selection',\n",
    "    'old_overdominance': 'Overdominance',\n",
    "    \n",
    "    'medium_selection': 'Selection',\n",
    "    'medium_balancingSelection': 'Balancing Selection',\n",
    "    'medium_negativeFreq-dep.Selection': 'Negative Freq-dep. Selection',\n",
    "    'medium_overdominance': 'Overdominance',\n",
    "}\n",
    "\n",
    "secure_blue = '#879BEF'\n",
    "peaceful_mint = '#FFBD00'\n",
    "innovation_purple = '#72bcd4'\n",
    "navy = '#3357A4'\n",
    "\n",
    "colors_dict = {\n",
    "    'old': [secure_blue, peaceful_mint, innovation_purple, navy],\n",
    "    'recent': [secure_blue, peaceful_mint, innovation_purple, navy],\n",
    "    'medium': [secure_blue, peaceful_mint, innovation_purple, navy]\n",
    "}\n",
    "\n",
    "for prefix in ['old', 'recent', 'medium']:\n",
    "    filename = f\"{prefix}_stacked_bar_plot.png\"\n",
    "    plot_stacked_bars(counts_df, prefix, legend_labels, colors_dict[prefix], filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available Dataframes:\n"
     ]
    }
   ],
   "source": [
    "available_dfs = [var_name for var_name, var in globals().items() if isinstance(var, pd.DataFrame)]\n",
    "\n",
    "print(\"Available Dataframes:\")\n",
    "# for df_name in available_dfs:\n",
    "#     print(df_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Archaic Matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>seg</th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>AltaiNean</th>\n",
       "      <th>AltaiDeni</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1594</th>\n",
       "      <td>22</td>\n",
       "      <td>103</td>\n",
       "      <td>49054049</td>\n",
       "      <td>49103380</td>\n",
       "      <td>0.5859</td>\n",
       "      <td>0.5636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1595</th>\n",
       "      <td>22</td>\n",
       "      <td>104</td>\n",
       "      <td>49103660</td>\n",
       "      <td>49153192</td>\n",
       "      <td>0.9831</td>\n",
       "      <td>0.9636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1596</th>\n",
       "      <td>22</td>\n",
       "      <td>105</td>\n",
       "      <td>49307031</td>\n",
       "      <td>49353474</td>\n",
       "      <td>0.1228</td>\n",
       "      <td>0.2586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1597</th>\n",
       "      <td>22</td>\n",
       "      <td>106</td>\n",
       "      <td>49404049</td>\n",
       "      <td>49450895</td>\n",
       "      <td>0.5225</td>\n",
       "      <td>0.5405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598</th>\n",
       "      <td>22</td>\n",
       "      <td>107</td>\n",
       "      <td>49460591</td>\n",
       "      <td>49502569</td>\n",
       "      <td>0.7818</td>\n",
       "      <td>0.5862</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr  seg      from        to  AltaiNean  AltaiDeni\n",
       "1594   22  103  49054049  49103380     0.5859     0.5636\n",
       "1595   22  104  49103660  49153192     0.9831     0.9636\n",
       "1596   22  105  49307031  49353474     0.1228     0.2586\n",
       "1597   22  106  49404049  49450895     0.5225     0.5405\n",
       "1598   22  107  49460591  49502569     0.7818     0.5862"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "archaic = pd.read_csv('/Users/kristinagrigaityte/PycharmProjects/pulls/Files/BSModel/Predictions/archaic.csv')\n",
    "archaic.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Positions</th>\n",
       "      <th>Pred</th>\n",
       "      <th>PredClass</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>123025</th>\n",
       "      <td>rs131718</td>\n",
       "      <td>49404789</td>\n",
       "      <td>0.586998</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123026</th>\n",
       "      <td>rs131715</td>\n",
       "      <td>49413787</td>\n",
       "      <td>0.585034</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123027</th>\n",
       "      <td>rs9616915</td>\n",
       "      <td>49464446</td>\n",
       "      <td>0.550545</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123028</th>\n",
       "      <td>rs739365</td>\n",
       "      <td>49487182</td>\n",
       "      <td>0.597457</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123029</th>\n",
       "      <td>rs6009951</td>\n",
       "      <td>49498216</td>\n",
       "      <td>0.574389</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNP  Positions      Pred                     PredClass  chr\n",
       "123025   rs131718   49404789  0.586998  Negative Freq-Dep. Selection   22\n",
       "123026   rs131715   49413787  0.585034  Negative Freq-Dep. Selection   22\n",
       "123027  rs9616915   49464446  0.550545  Negative Freq-Dep. Selection   22\n",
       "123028   rs739365   49487182  0.597457  Negative Freq-Dep. Selection   22\n",
       "123029  rs6009951   49498216  0.574389  Negative Freq-Dep. Selection   22"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def add_chr_column(df, chr_number):\n",
    "    df = df.copy()\n",
    "    df.loc[:, 'chr'] = chr_number\n",
    "    return df\n",
    "\n",
    "def get_df_name(prefix, chr_number):\n",
    "    return f'{prefix}_ch{chr_number}'\n",
    "\n",
    "prefixes = ['prediction_recent_1', 'prediction_recent_2', 'prediction_recent_3']\n",
    "pred_classes = [\"Selection\", \"Balancing Selection\", \"Overdominance\", \"Negative Freq-Dep. Selection\"]\n",
    "all_dfs = []\n",
    "\n",
    "for chr_number in range(1, 23):   \n",
    "    for prefix in prefixes:\n",
    "        df_name = get_df_name(prefix, chr_number)\n",
    "        try:\n",
    "            df = globals()[df_name]\n",
    "            for pred_class in pred_classes:\n",
    "                filtered_df = df[df['PredClass'] == pred_class]\n",
    "                if not filtered_df.empty:\n",
    "                    filtered_df = add_chr_column(filtered_df, chr_number)\n",
    "                    all_dfs.append(filtered_df)\n",
    "        except KeyError:\n",
    "            continue\n",
    "\n",
    "all_data = pd.concat(all_dfs, ignore_index=True)\n",
    "all_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Positions</th>\n",
       "      <th>Pred</th>\n",
       "      <th>PredClass</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>rs10927011</td>\n",
       "      <td>241645735</td>\n",
       "      <td>0.550709</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>rs692981</td>\n",
       "      <td>242232230</td>\n",
       "      <td>0.517401</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>rs1000543</td>\n",
       "      <td>242254223</td>\n",
       "      <td>0.581432</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>rs10927137</td>\n",
       "      <td>242355096</td>\n",
       "      <td>0.580353</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>rs913994</td>\n",
       "      <td>242381803</td>\n",
       "      <td>0.574044</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            SNP  Positions      Pred            PredClass  chr\n",
       "150  rs10927011  241645735  0.550709  Balancing Selection    1\n",
       "151    rs692981  242232230  0.517401  Balancing Selection    1\n",
       "152   rs1000543  242254223  0.581432  Balancing Selection    1\n",
       "153  rs10927137  242355096  0.580353  Balancing Selection    1\n",
       "154    rs913994  242381803  0.574044  Balancing Selection    1"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selection_snps = all_data[all_data['PredClass'] == 'Selection']['SNP'].unique()\n",
    "filtered_df = all_data[all_data['PredClass'].isin(['Balancing Selection', 'Overdominance', 'Negative Freq-Dep. Selection'])]\n",
    "all_data = filtered_df[filtered_df['SNP'].isin(selection_snps)]\n",
    "all_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 892 ms, sys: 3.86 ms, total: 896 ms\n",
      "Wall time: 903 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "def find_snps_in_range(row, all_data):\n",
    "    snps_in_range = all_data[\n",
    "        (all_data['chr'] == row['chr']) & \n",
    "        (all_data['Positions'] >= row['from']) & \n",
    "        (all_data['Positions'] <= row['to'])\n",
    "    ]\n",
    "    grouped_snps = snps_in_range.groupby('PredClass').agg({'SNP': list}).reset_index()\n",
    "    return grouped_snps\n",
    "\n",
    "new_rows = []\n",
    "\n",
    "for _, row in archaic.iterrows():\n",
    "    snps_grouped = find_snps_in_range(row, all_data)\n",
    "    for _, snp_group in snps_grouped.iterrows():\n",
    "        new_rows.append([\n",
    "            row['chr'], row['seg'], row['from'], row['to'], row['AltaiNean'], row['AltaiDeni'],\n",
    "            snp_group['SNP'], snp_group['PredClass']\n",
    "        ])\n",
    "\n",
    "flattened_df = pd.DataFrame(new_rows, columns=['chr', 'seg', 'from', 'to', 'AltaiNean', 'AltaiDeni', 'SNPs', 'PredClass'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>seg</th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>AltaiNean</th>\n",
       "      <th>AltaiDeni</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>PredClass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>22.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>49003874.0</td>\n",
       "      <td>49053529.0</td>\n",
       "      <td>0.1111</td>\n",
       "      <td>0.4744</td>\n",
       "      <td>[rs742186, rs2235356]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>22.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>49307031.0</td>\n",
       "      <td>49353474.0</td>\n",
       "      <td>0.1228</td>\n",
       "      <td>0.2586</td>\n",
       "      <td>[rs140519]</td>\n",
       "      <td>Balancing Selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>22.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>49307031.0</td>\n",
       "      <td>49353474.0</td>\n",
       "      <td>0.1228</td>\n",
       "      <td>0.2586</td>\n",
       "      <td>[rs140519, rs131778]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>22.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>49404049.0</td>\n",
       "      <td>49450895.0</td>\n",
       "      <td>0.5225</td>\n",
       "      <td>0.5405</td>\n",
       "      <td>[rs131718, rs131715]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>22.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>49460591.0</td>\n",
       "      <td>49502569.0</td>\n",
       "      <td>0.7818</td>\n",
       "      <td>0.5862</td>\n",
       "      <td>[rs9616915, rs739365, rs6009951]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    seg        from          to  AltaiNean  AltaiDeni  \\\n",
       "171  22.0  102.0  49003874.0  49053529.0     0.1111     0.4744   \n",
       "172  22.0  105.0  49307031.0  49353474.0     0.1228     0.2586   \n",
       "173  22.0  105.0  49307031.0  49353474.0     0.1228     0.2586   \n",
       "174  22.0  106.0  49404049.0  49450895.0     0.5225     0.5405   \n",
       "175  22.0  107.0  49460591.0  49502569.0     0.7818     0.5862   \n",
       "\n",
       "                                 SNPs                     PredClass  \n",
       "171             [rs742186, rs2235356]  Negative Freq-Dep. Selection  \n",
       "172                        [rs140519]           Balancing Selection  \n",
       "173              [rs140519, rs131778]  Negative Freq-Dep. Selection  \n",
       "174              [rs131718, rs131715]  Negative Freq-Dep. Selection  \n",
       "175  [rs9616915, rs739365, rs6009951]  Negative Freq-Dep. Selection  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flattened_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>seg</th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>AltaiNean</th>\n",
       "      <th>AltaiDeni</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>PredClass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>22.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>35654133.0</td>\n",
       "      <td>35703551.0</td>\n",
       "      <td>0.8596</td>\n",
       "      <td>0.8305</td>\n",
       "      <td>[rs11089816, rs5750348]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>22.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>36103709.0</td>\n",
       "      <td>36133397.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.8636</td>\n",
       "      <td>[rs4820286, rs5750428]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>22.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>44107968.0</td>\n",
       "      <td>44149935.0</td>\n",
       "      <td>0.9167</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>[rs11704481, rs2064068, rs1044742, rs5764698]</td>\n",
       "      <td>Balancing Selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>22.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>44107968.0</td>\n",
       "      <td>44149935.0</td>\n",
       "      <td>0.9167</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>[rs11704481, rs2064068, rs1044742, rs5764698]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>22.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>47432419.0</td>\n",
       "      <td>47445957.0</td>\n",
       "      <td>0.9107</td>\n",
       "      <td>0.9630</td>\n",
       "      <td>[rs5771906, rs6010568]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr   seg        from          to  AltaiNean  AltaiDeni  \\\n",
       "132  22.0  56.0  35654133.0  35703551.0     0.8596     0.8305   \n",
       "135  22.0  59.0  36103709.0  36133397.0     1.0000     0.8636   \n",
       "146  22.0  81.0  44107968.0  44149935.0     0.9167     1.0000   \n",
       "147  22.0  81.0  44107968.0  44149935.0     0.9167     1.0000   \n",
       "160  22.0  93.0  47432419.0  47445957.0     0.9107     0.9630   \n",
       "\n",
       "                                              SNPs  \\\n",
       "132                        [rs11089816, rs5750348]   \n",
       "135                         [rs4820286, rs5750428]   \n",
       "146  [rs11704481, rs2064068, rs1044742, rs5764698]   \n",
       "147  [rs11704481, rs2064068, rs1044742, rs5764698]   \n",
       "160                         [rs5771906, rs6010568]   \n",
       "\n",
       "                        PredClass  \n",
       "132  Negative Freq-Dep. Selection  \n",
       "135  Negative Freq-Dep. Selection  \n",
       "146           Balancing Selection  \n",
       "147  Negative Freq-Dep. Selection  \n",
       "160  Negative Freq-Dep. Selection  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df = flattened_df[(flattened_df['AltaiNean'] > 0.85)]\n",
    "filtered_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "def get_genes_and_consequence_for_snp(snp_id, chr_column):\n",
    "    server = \"https://grch37.rest.ensembl.org\"\n",
    "    ext = f\"/variation/human/{snp_id}?content-type=application/json\"\n",
    "    r = requests.get(server + ext, headers={\"Content-Type\": \"application/json\"})\n",
    "    \n",
    "    if not r.ok:\n",
    "        return [], None\n",
    "    \n",
    "    data = r.json()\n",
    "    genes = []\n",
    "    consequence = data.get(\"most_severe_consequence\", \"unknown\")\n",
    "    \n",
    "    if 'mappings' in data:\n",
    "        for mapping in data['mappings']:\n",
    "            location = f\"{chr_column}:{mapping['start']}-{mapping['end']}\"\n",
    "            ext = f\"/overlap/region/human/{location}?feature=gene;content-type=application/json\"\n",
    "            r = requests.get(server + ext, headers={\"Content-Type\": \"application/json\"})\n",
    "            \n",
    "            if not r.ok:\n",
    "                continue\n",
    "            \n",
    "            gene_data = r.json()\n",
    "            \n",
    "            for gene in gene_data:\n",
    "                genes.append(gene['id'])   \n",
    "    \n",
    "    return list(set(genes)), consequence\n",
    "\n",
    "def add_gene_info(row):\n",
    "    all_genes = []\n",
    "    consequences = []\n",
    "    chr_column = str(int(row['chr']))  \n",
    "    \n",
    "    for snp in row['SNPs']:\n",
    "        gene_ids, consequence = get_genes_and_consequence_for_snp(snp, chr_column)\n",
    "        if gene_ids:\n",
    "            all_genes.extend(gene_ids)\n",
    "        consequences.append(consequence)\n",
    "    \n",
    "    return list(set(all_genes)), consequences\n",
    "\n",
    "filtered_df = filtered_df.copy()\n",
    "filtered_df[['Genes', 'Consequences']] = filtered_df.apply(\n",
    "    lambda row: pd.Series(add_gene_info(row)), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>seg</th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>AltaiNean</th>\n",
       "      <th>AltaiDeni</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>PredClass</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Consequences</th>\n",
       "      <th>Proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>189261797.0</td>\n",
       "      <td>189309846.0</td>\n",
       "      <td>0.9394</td>\n",
       "      <td>0.8154</td>\n",
       "      <td>[rs2037184, rs6444269]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>191112052.0</td>\n",
       "      <td>191161035.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.8298</td>\n",
       "      <td>[rs12696598, rs724438, rs4687112]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, regulatory_region_variant, regulatory_region_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>191661171.0</td>\n",
       "      <td>191710630.0</td>\n",
       "      <td>0.9479</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>[rs3943979, rs9857115, rs2042865]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, intergenic_variant, regulatory_region_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>191711064.0</td>\n",
       "      <td>191755302.0</td>\n",
       "      <td>0.9818</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>[rs2193880, rs3773989]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000196083]</td>\n",
       "      <td>[intron_variant, intron_variant]</td>\n",
       "      <td>[[IL1RAP, interleukin 1 receptor accessory protein [Source:HGNC Symbol;Acc:5995]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>186117497.0</td>\n",
       "      <td>186164767.0</td>\n",
       "      <td>0.9268</td>\n",
       "      <td>0.7750</td>\n",
       "      <td>[rs4355430]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr    seg         from           to  AltaiNean  AltaiDeni  \\\n",
       "9   3.0   98.0  189261797.0  189309846.0     0.9394     0.8154   \n",
       "10  3.0  100.0  191112052.0  191161035.0     1.0000     0.8298   \n",
       "12  3.0  103.0  191661171.0  191710630.0     0.9479     0.9500   \n",
       "13  3.0  104.0  191711064.0  191755302.0     0.9818     1.0000   \n",
       "15  4.0  141.0  186117497.0  186164767.0     0.9268     0.7750   \n",
       "\n",
       "                                 SNPs                     PredClass  \\\n",
       "9              [rs2037184, rs6444269]  Negative Freq-Dep. Selection   \n",
       "10  [rs12696598, rs724438, rs4687112]  Negative Freq-Dep. Selection   \n",
       "12  [rs3943979, rs9857115, rs2042865]  Negative Freq-Dep. Selection   \n",
       "13             [rs2193880, rs3773989]  Negative Freq-Dep. Selection   \n",
       "15                        [rs4355430]  Negative Freq-Dep. Selection   \n",
       "\n",
       "                Genes  \\\n",
       "9                  []   \n",
       "10                 []   \n",
       "12                 []   \n",
       "13  [ENSG00000196083]   \n",
       "15                 []   \n",
       "\n",
       "                                                                  Consequences  \\\n",
       "9                                     [intergenic_variant, intergenic_variant]   \n",
       "10  [intergenic_variant, regulatory_region_variant, regulatory_region_variant]   \n",
       "12         [intergenic_variant, intergenic_variant, regulatory_region_variant]   \n",
       "13                                            [intron_variant, intron_variant]   \n",
       "15                                                        [intergenic_variant]   \n",
       "\n",
       "                                                                              Proteins  \n",
       "9                                                                                   []  \n",
       "10                                                                                  []  \n",
       "12                                                                                  []  \n",
       "13  [[IL1RAP, interleukin 1 receptor accessory protein [Source:HGNC Symbol;Acc:5995]]]  \n",
       "15                                                                                  []  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.21 s, sys: 56.6 ms, total: 1.26 s\n",
      "Wall time: 6.34 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "def get_protein_info_for_gene(gene_id):\n",
    "    server = \"https://grch37.rest.ensembl.org\"\n",
    "    ext = f\"/lookup/id/{gene_id}?content-type=application/json\"\n",
    "    r = requests.get(server + ext, headers={\"Content-Type\": \"application/json\"})\n",
    "    \n",
    "    if not r.ok:\n",
    "        return None\n",
    "    \n",
    "    data = r.json()\n",
    "    protein_name = data.get('display_name', 'Unknown')\n",
    "    description = data.get('description', 'No description available')\n",
    "    \n",
    "    return [protein_name, description]\n",
    "\n",
    "def add_protein_info(genes):\n",
    "    protein_info = []\n",
    "    for gene in genes:\n",
    "        info = get_protein_info_for_gene(gene)\n",
    "        if info:\n",
    "            protein_info.append(info)\n",
    "    return protein_info\n",
    "\n",
    "filtered_df['Proteins'] = filtered_df['Genes'].apply(add_protein_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>seg</th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>AltaiNean</th>\n",
       "      <th>AltaiDeni</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>PredClass</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Consequences</th>\n",
       "      <th>Proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>189261797.0</td>\n",
       "      <td>189309846.0</td>\n",
       "      <td>0.9394</td>\n",
       "      <td>0.8154</td>\n",
       "      <td>[rs2037184, rs6444269]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>191112052.0</td>\n",
       "      <td>191161035.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.8298</td>\n",
       "      <td>[rs12696598, rs724438, rs4687112]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, regulatory_region_variant, regulatory_region_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>191661171.0</td>\n",
       "      <td>191710630.0</td>\n",
       "      <td>0.9479</td>\n",
       "      <td>0.9500</td>\n",
       "      <td>[rs3943979, rs9857115, rs2042865]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, intergenic_variant, regulatory_region_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>191711064.0</td>\n",
       "      <td>191755302.0</td>\n",
       "      <td>0.9818</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>[rs2193880, rs3773989]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000196083]</td>\n",
       "      <td>[intron_variant, intron_variant]</td>\n",
       "      <td>[[IL1RAP, interleukin 1 receptor accessory protein [Source:HGNC Symbol;Acc:5995]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>186117497.0</td>\n",
       "      <td>186164767.0</td>\n",
       "      <td>0.9268</td>\n",
       "      <td>0.7750</td>\n",
       "      <td>[rs4355430]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>6.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>143867336.0</td>\n",
       "      <td>143897977.0</td>\n",
       "      <td>0.9661</td>\n",
       "      <td>0.1379</td>\n",
       "      <td>[rs2128977]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[regulatory_region_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>6.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>156898474.0</td>\n",
       "      <td>156946230.0</td>\n",
       "      <td>0.8947</td>\n",
       "      <td>0.9459</td>\n",
       "      <td>[rs288976]</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>6.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>156898474.0</td>\n",
       "      <td>156946230.0</td>\n",
       "      <td>0.8947</td>\n",
       "      <td>0.9459</td>\n",
       "      <td>[rs288976]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>6.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>159198260.0</td>\n",
       "      <td>159247967.0</td>\n",
       "      <td>0.9672</td>\n",
       "      <td>0.2143</td>\n",
       "      <td>[rs9457507, rs9457511, rs12200537]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000203711]</td>\n",
       "      <td>[intron_variant, intron_variant, intron_variant]</td>\n",
       "      <td>[[C6orf99, chromosome 6 open reading frame 99 [Source:HGNC Symbol;Acc:21179]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>6.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>161248552.0</td>\n",
       "      <td>161291617.0</td>\n",
       "      <td>0.9286</td>\n",
       "      <td>0.9032</td>\n",
       "      <td>[rs1962358]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>6.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>163048514.0</td>\n",
       "      <td>163097724.0</td>\n",
       "      <td>0.9091</td>\n",
       "      <td>0.4182</td>\n",
       "      <td>[rs9356058]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000112530]</td>\n",
       "      <td>[intron_variant]</td>\n",
       "      <td>[[PACRG, PARK2 co-regulated [Source:HGNC Symbol;Acc:19152]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>9.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>129660977.0</td>\n",
       "      <td>129709828.0</td>\n",
       "      <td>0.8684</td>\n",
       "      <td>0.7714</td>\n",
       "      <td>[rs3780663]</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>[ENSG00000257524, ENSG00000106992]</td>\n",
       "      <td>[intron_variant]</td>\n",
       "      <td>[[RP11-203J24.9, No description available], [AK1, adenylate kinase 1 [Source:HGNC Symbol;Acc:361]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>9.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>129660977.0</td>\n",
       "      <td>129709828.0</td>\n",
       "      <td>0.8684</td>\n",
       "      <td>0.7714</td>\n",
       "      <td>[rs3780663]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000257524, ENSG00000106992]</td>\n",
       "      <td>[intron_variant]</td>\n",
       "      <td>[[RP11-203J24.9, No description available], [AK1, adenylate kinase 1 [Source:HGNC Symbol;Acc:361]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>10.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>127671175.0</td>\n",
       "      <td>127711500.0</td>\n",
       "      <td>0.9787</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>[rs12251249]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000148848]</td>\n",
       "      <td>[intron_variant]</td>\n",
       "      <td>[[ADAM12, ADAM metallopeptidase domain 12 [Source:HGNC Symbol;Acc:190]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>13.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>102071812.0</td>\n",
       "      <td>102116884.0</td>\n",
       "      <td>0.8611</td>\n",
       "      <td>0.8734</td>\n",
       "      <td>[rs3736972]</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>[ENSG00000134900]</td>\n",
       "      <td>[synonymous_variant]</td>\n",
       "      <td>[[TPP2, tripeptidyl peptidase II [Source:HGNC Symbol;Acc:12016]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>13.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>102071812.0</td>\n",
       "      <td>102116884.0</td>\n",
       "      <td>0.8611</td>\n",
       "      <td>0.8734</td>\n",
       "      <td>[rs3736972]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000134900]</td>\n",
       "      <td>[synonymous_variant]</td>\n",
       "      <td>[[TPP2, tripeptidyl peptidase II [Source:HGNC Symbol;Acc:12016]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>13.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>107421329.0</td>\n",
       "      <td>107469601.0</td>\n",
       "      <td>0.9365</td>\n",
       "      <td>0.6349</td>\n",
       "      <td>[rs9520690, rs9520691, rs9520696, rs9514782, rs9514792, rs12871532, rs1022909]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[regulatory_region_variant, intergenic_variant, intergenic_variant, intergenic_variant, intergenic_variant, intergenic_variant, intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>15.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>86002672.0</td>\n",
       "      <td>86049988.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.9483</td>\n",
       "      <td>[rs6496435]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000259560]</td>\n",
       "      <td>[intron_variant]</td>\n",
       "      <td>[[RP11-648K4.2, No description available]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>15.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>89501124.0</td>\n",
       "      <td>89544688.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>[rs12905479, rs17594552]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000185518]</td>\n",
       "      <td>[intron_variant, intron_variant]</td>\n",
       "      <td>[[SV2B, synaptic vesicle glycoprotein 2B [Source:HGNC Symbol;Acc:16874]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>17.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>65702659.0</td>\n",
       "      <td>65749953.0</td>\n",
       "      <td>0.8519</td>\n",
       "      <td>0.6786</td>\n",
       "      <td>[rs236586, rs236523, rs236531]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, intergenic_variant, intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>17.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>68256221.0</td>\n",
       "      <td>68299201.0</td>\n",
       "      <td>0.9333</td>\n",
       "      <td>0.9839</td>\n",
       "      <td>[rs1110614, rs4793484, rs9906450]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000133195]</td>\n",
       "      <td>[intron_variant, intron_variant, intron_variant]</td>\n",
       "      <td>[[SLC39A11, solute carrier family 39, member 11 [Source:HGNC Symbol;Acc:14463]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>19.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>52928986.0</td>\n",
       "      <td>52976774.0</td>\n",
       "      <td>0.8989</td>\n",
       "      <td>0.8452</td>\n",
       "      <td>[rs2334295]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000269656]</td>\n",
       "      <td>[intron_variant]</td>\n",
       "      <td>[[CTD-2571L23.6, No description available]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>21.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>28411654.0</td>\n",
       "      <td>28456611.0</td>\n",
       "      <td>0.8571</td>\n",
       "      <td>0.8636</td>\n",
       "      <td>[rs2223743, rs6516819, rs9976944, rs2831534]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000236532, ENSG00000232079]</td>\n",
       "      <td>[intron_variant, intron_variant, intron_variant, intron_variant]</td>\n",
       "      <td>[[AL035610.2, No description available], [AL035610.1, No description available]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>22.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>34153779.0</td>\n",
       "      <td>34203496.0</td>\n",
       "      <td>0.9706</td>\n",
       "      <td>0.7778</td>\n",
       "      <td>[rs5755730]</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>22.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>34153779.0</td>\n",
       "      <td>34203496.0</td>\n",
       "      <td>0.9706</td>\n",
       "      <td>0.7778</td>\n",
       "      <td>[rs5755730]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>22.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>35654133.0</td>\n",
       "      <td>35703551.0</td>\n",
       "      <td>0.8596</td>\n",
       "      <td>0.8305</td>\n",
       "      <td>[rs11089816, rs5750348]</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>22.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>35654133.0</td>\n",
       "      <td>35703551.0</td>\n",
       "      <td>0.8596</td>\n",
       "      <td>0.8305</td>\n",
       "      <td>[rs11089816, rs5750348]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant, intergenic_variant]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>22.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>36103709.0</td>\n",
       "      <td>36133397.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.8636</td>\n",
       "      <td>[rs4820286, rs5750428]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000243902, ENSG00000166897]</td>\n",
       "      <td>[intron_variant, intron_variant]</td>\n",
       "      <td>[[RP1-63G5.5, No description available], [ELFN2, extracellular leucine-rich repeat and fibronectin type III domain containing 2 [Source:HGNC Symbol;Acc:29396]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>22.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>44107968.0</td>\n",
       "      <td>44149935.0</td>\n",
       "      <td>0.9167</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>[rs11704481, rs2064068, rs1044742, rs5764698]</td>\n",
       "      <td>Balancing Selection</td>\n",
       "      <td>[ENSG00000100376, ENSG00000077935]</td>\n",
       "      <td>[intron_variant, 3_prime_UTR_variant, 3_prime_UTR_variant, missense_variant]</td>\n",
       "      <td>[[FAM118A, family with sequence similarity 118, member A [Source:HGNC Symbol;Acc:1313]], [SMC1B, structural maintenance of chromosomes 1B [Source:HGNC Symbol;Acc:11112]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>22.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>44107968.0</td>\n",
       "      <td>44149935.0</td>\n",
       "      <td>0.9167</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>[rs11704481, rs2064068, rs1044742, rs5764698]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000100376, ENSG00000077935]</td>\n",
       "      <td>[intron_variant, 3_prime_UTR_variant, 3_prime_UTR_variant, missense_variant]</td>\n",
       "      <td>[[FAM118A, family with sequence similarity 118, member A [Source:HGNC Symbol;Acc:1313]], [SMC1B, structural maintenance of chromosomes 1B [Source:HGNC Symbol;Acc:11112]]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>22.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>47432419.0</td>\n",
       "      <td>47445957.0</td>\n",
       "      <td>0.9107</td>\n",
       "      <td>0.9630</td>\n",
       "      <td>[rs5771906, rs6010568]</td>\n",
       "      <td>Negative Freq-Dep. Selection</td>\n",
       "      <td>[ENSG00000219438]</td>\n",
       "      <td>[intron_variant, intron_variant]</td>\n",
       "      <td>[[FAM19A5, family with sequence similarity 19 (chemokine (C-C motif)-like), member A5 [Source:HGNC Symbol;Acc:21592]]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    seg         from           to  AltaiNean  AltaiDeni  \\\n",
       "9     3.0   98.0  189261797.0  189309846.0     0.9394     0.8154   \n",
       "10    3.0  100.0  191112052.0  191161035.0     1.0000     0.8298   \n",
       "12    3.0  103.0  191661171.0  191710630.0     0.9479     0.9500   \n",
       "13    3.0  104.0  191711064.0  191755302.0     0.9818     1.0000   \n",
       "15    4.0  141.0  186117497.0  186164767.0     0.9268     0.7750   \n",
       "17    6.0  104.0  143867336.0  143897977.0     0.9661     0.1379   \n",
       "18    6.0  110.0  156898474.0  156946230.0     0.8947     0.9459   \n",
       "19    6.0  110.0  156898474.0  156946230.0     0.8947     0.9459   \n",
       "20    6.0  112.0  159198260.0  159247967.0     0.9672     0.2143   \n",
       "23    6.0  115.0  161248552.0  161291617.0     0.9286     0.9032   \n",
       "25    6.0  117.0  163048514.0  163097724.0     0.9091     0.4182   \n",
       "31    9.0   65.0  129660977.0  129709828.0     0.8684     0.7714   \n",
       "32    9.0   65.0  129660977.0  129709828.0     0.8684     0.7714   \n",
       "33   10.0   80.0  127671175.0  127711500.0     0.9787     1.0000   \n",
       "48   13.0   34.0  102071812.0  102116884.0     0.8611     0.8734   \n",
       "49   13.0   34.0  102071812.0  102116884.0     0.8611     0.8734   \n",
       "52   13.0   39.0  107421329.0  107469601.0     0.9365     0.6349   \n",
       "53   15.0   14.0   86002672.0   86049988.0     1.0000     0.9483   \n",
       "54   15.0   16.0   89501124.0   89544688.0     1.0000     1.0000   \n",
       "68   17.0   19.0   65702659.0   65749953.0     0.8519     0.6786   \n",
       "71   17.0   22.0   68256221.0   68299201.0     0.9333     0.9839   \n",
       "86   19.0   34.0   52928986.0   52976774.0     0.8989     0.8452   \n",
       "92   21.0   27.0   28411654.0   28456611.0     0.8571     0.8636   \n",
       "122  22.0   46.0   34153779.0   34203496.0     0.9706     0.7778   \n",
       "123  22.0   46.0   34153779.0   34203496.0     0.9706     0.7778   \n",
       "131  22.0   56.0   35654133.0   35703551.0     0.8596     0.8305   \n",
       "132  22.0   56.0   35654133.0   35703551.0     0.8596     0.8305   \n",
       "135  22.0   59.0   36103709.0   36133397.0     1.0000     0.8636   \n",
       "146  22.0   81.0   44107968.0   44149935.0     0.9167     1.0000   \n",
       "147  22.0   81.0   44107968.0   44149935.0     0.9167     1.0000   \n",
       "160  22.0   93.0   47432419.0   47445957.0     0.9107     0.9630   \n",
       "\n",
       "                                                                               SNPs  \\\n",
       "9                                                            [rs2037184, rs6444269]   \n",
       "10                                                [rs12696598, rs724438, rs4687112]   \n",
       "12                                                [rs3943979, rs9857115, rs2042865]   \n",
       "13                                                           [rs2193880, rs3773989]   \n",
       "15                                                                      [rs4355430]   \n",
       "17                                                                      [rs2128977]   \n",
       "18                                                                       [rs288976]   \n",
       "19                                                                       [rs288976]   \n",
       "20                                               [rs9457507, rs9457511, rs12200537]   \n",
       "23                                                                      [rs1962358]   \n",
       "25                                                                      [rs9356058]   \n",
       "31                                                                      [rs3780663]   \n",
       "32                                                                      [rs3780663]   \n",
       "33                                                                     [rs12251249]   \n",
       "48                                                                      [rs3736972]   \n",
       "49                                                                      [rs3736972]   \n",
       "52   [rs9520690, rs9520691, rs9520696, rs9514782, rs9514792, rs12871532, rs1022909]   \n",
       "53                                                                      [rs6496435]   \n",
       "54                                                         [rs12905479, rs17594552]   \n",
       "68                                                   [rs236586, rs236523, rs236531]   \n",
       "71                                                [rs1110614, rs4793484, rs9906450]   \n",
       "86                                                                      [rs2334295]   \n",
       "92                                     [rs2223743, rs6516819, rs9976944, rs2831534]   \n",
       "122                                                                     [rs5755730]   \n",
       "123                                                                     [rs5755730]   \n",
       "131                                                         [rs11089816, rs5750348]   \n",
       "132                                                         [rs11089816, rs5750348]   \n",
       "135                                                          [rs4820286, rs5750428]   \n",
       "146                                   [rs11704481, rs2064068, rs1044742, rs5764698]   \n",
       "147                                   [rs11704481, rs2064068, rs1044742, rs5764698]   \n",
       "160                                                          [rs5771906, rs6010568]   \n",
       "\n",
       "                        PredClass                               Genes  \\\n",
       "9    Negative Freq-Dep. Selection                                  []   \n",
       "10   Negative Freq-Dep. Selection                                  []   \n",
       "12   Negative Freq-Dep. Selection                                  []   \n",
       "13   Negative Freq-Dep. Selection                   [ENSG00000196083]   \n",
       "15   Negative Freq-Dep. Selection                                  []   \n",
       "17   Negative Freq-Dep. Selection                                  []   \n",
       "18            Balancing Selection                                  []   \n",
       "19   Negative Freq-Dep. Selection                                  []   \n",
       "20   Negative Freq-Dep. Selection                   [ENSG00000203711]   \n",
       "23   Negative Freq-Dep. Selection                                  []   \n",
       "25   Negative Freq-Dep. Selection                   [ENSG00000112530]   \n",
       "31            Balancing Selection  [ENSG00000257524, ENSG00000106992]   \n",
       "32   Negative Freq-Dep. Selection  [ENSG00000257524, ENSG00000106992]   \n",
       "33   Negative Freq-Dep. Selection                   [ENSG00000148848]   \n",
       "48            Balancing Selection                   [ENSG00000134900]   \n",
       "49   Negative Freq-Dep. Selection                   [ENSG00000134900]   \n",
       "52   Negative Freq-Dep. Selection                                  []   \n",
       "53   Negative Freq-Dep. Selection                   [ENSG00000259560]   \n",
       "54   Negative Freq-Dep. Selection                   [ENSG00000185518]   \n",
       "68   Negative Freq-Dep. Selection                                  []   \n",
       "71   Negative Freq-Dep. Selection                   [ENSG00000133195]   \n",
       "86   Negative Freq-Dep. Selection                   [ENSG00000269656]   \n",
       "92   Negative Freq-Dep. Selection  [ENSG00000236532, ENSG00000232079]   \n",
       "122           Balancing Selection                                  []   \n",
       "123  Negative Freq-Dep. Selection                                  []   \n",
       "131           Balancing Selection                                  []   \n",
       "132  Negative Freq-Dep. Selection                                  []   \n",
       "135  Negative Freq-Dep. Selection  [ENSG00000243902, ENSG00000166897]   \n",
       "146           Balancing Selection  [ENSG00000100376, ENSG00000077935]   \n",
       "147  Negative Freq-Dep. Selection  [ENSG00000100376, ENSG00000077935]   \n",
       "160  Negative Freq-Dep. Selection                   [ENSG00000219438]   \n",
       "\n",
       "                                                                                                                                            Consequences  \\\n",
       "9                                                                                                               [intergenic_variant, intergenic_variant]   \n",
       "10                                                                            [intergenic_variant, regulatory_region_variant, regulatory_region_variant]   \n",
       "12                                                                                   [intergenic_variant, intergenic_variant, regulatory_region_variant]   \n",
       "13                                                                                                                      [intron_variant, intron_variant]   \n",
       "15                                                                                                                                  [intergenic_variant]   \n",
       "17                                                                                                                           [regulatory_region_variant]   \n",
       "18                                                                                                                                  [intergenic_variant]   \n",
       "19                                                                                                                                  [intergenic_variant]   \n",
       "20                                                                                                      [intron_variant, intron_variant, intron_variant]   \n",
       "23                                                                                                                                  [intergenic_variant]   \n",
       "25                                                                                                                                      [intron_variant]   \n",
       "31                                                                                                                                      [intron_variant]   \n",
       "32                                                                                                                                      [intron_variant]   \n",
       "33                                                                                                                                      [intron_variant]   \n",
       "48                                                                                                                                  [synonymous_variant]   \n",
       "49                                                                                                                                  [synonymous_variant]   \n",
       "52   [regulatory_region_variant, intergenic_variant, intergenic_variant, intergenic_variant, intergenic_variant, intergenic_variant, intergenic_variant]   \n",
       "53                                                                                                                                      [intron_variant]   \n",
       "54                                                                                                                      [intron_variant, intron_variant]   \n",
       "68                                                                                          [intergenic_variant, intergenic_variant, intergenic_variant]   \n",
       "71                                                                                                      [intron_variant, intron_variant, intron_variant]   \n",
       "86                                                                                                                                      [intron_variant]   \n",
       "92                                                                                      [intron_variant, intron_variant, intron_variant, intron_variant]   \n",
       "122                                                                                                                                 [intergenic_variant]   \n",
       "123                                                                                                                                 [intergenic_variant]   \n",
       "131                                                                                                             [intergenic_variant, intergenic_variant]   \n",
       "132                                                                                                             [intergenic_variant, intergenic_variant]   \n",
       "135                                                                                                                     [intron_variant, intron_variant]   \n",
       "146                                                                         [intron_variant, 3_prime_UTR_variant, 3_prime_UTR_variant, missense_variant]   \n",
       "147                                                                         [intron_variant, 3_prime_UTR_variant, 3_prime_UTR_variant, missense_variant]   \n",
       "160                                                                                                                     [intron_variant, intron_variant]   \n",
       "\n",
       "                                                                                                                                                                       Proteins  \n",
       "9                                                                                                                                                                            []  \n",
       "10                                                                                                                                                                           []  \n",
       "12                                                                                                                                                                           []  \n",
       "13                                                                                           [[IL1RAP, interleukin 1 receptor accessory protein [Source:HGNC Symbol;Acc:5995]]]  \n",
       "15                                                                                                                                                                           []  \n",
       "17                                                                                                                                                                           []  \n",
       "18                                                                                                                                                                           []  \n",
       "19                                                                                                                                                                           []  \n",
       "20                                                                                               [[C6orf99, chromosome 6 open reading frame 99 [Source:HGNC Symbol;Acc:21179]]]  \n",
       "23                                                                                                                                                                           []  \n",
       "25                                                                                                                 [[PACRG, PARK2 co-regulated [Source:HGNC Symbol;Acc:19152]]]  \n",
       "31                                                                          [[RP11-203J24.9, No description available], [AK1, adenylate kinase 1 [Source:HGNC Symbol;Acc:361]]]  \n",
       "32                                                                          [[RP11-203J24.9, No description available], [AK1, adenylate kinase 1 [Source:HGNC Symbol;Acc:361]]]  \n",
       "33                                                                                                     [[ADAM12, ADAM metallopeptidase domain 12 [Source:HGNC Symbol;Acc:190]]]  \n",
       "48                                                                                                            [[TPP2, tripeptidyl peptidase II [Source:HGNC Symbol;Acc:12016]]]  \n",
       "49                                                                                                            [[TPP2, tripeptidyl peptidase II [Source:HGNC Symbol;Acc:12016]]]  \n",
       "52                                                                                                                                                                           []  \n",
       "53                                                                                                                                   [[RP11-648K4.2, No description available]]  \n",
       "54                                                                                                    [[SV2B, synaptic vesicle glycoprotein 2B [Source:HGNC Symbol;Acc:16874]]]  \n",
       "68                                                                                                                                                                           []  \n",
       "71                                                                                             [[SLC39A11, solute carrier family 39, member 11 [Source:HGNC Symbol;Acc:14463]]]  \n",
       "86                                                                                                                                  [[CTD-2571L23.6, No description available]]  \n",
       "92                                                                                             [[AL035610.2, No description available], [AL035610.1, No description available]]  \n",
       "122                                                                                                                                                                          []  \n",
       "123                                                                                                                                                                          []  \n",
       "131                                                                                                                                                                          []  \n",
       "132                                                                                                                                                                          []  \n",
       "135            [[RP1-63G5.5, No description available], [ELFN2, extracellular leucine-rich repeat and fibronectin type III domain containing 2 [Source:HGNC Symbol;Acc:29396]]]  \n",
       "146  [[FAM118A, family with sequence similarity 118, member A [Source:HGNC Symbol;Acc:1313]], [SMC1B, structural maintenance of chromosomes 1B [Source:HGNC Symbol;Acc:11112]]]  \n",
       "147  [[FAM118A, family with sequence similarity 118, member A [Source:HGNC Symbol;Acc:1313]], [SMC1B, structural maintenance of chromosomes 1B [Source:HGNC Symbol;Acc:11112]]]  \n",
       "160                                                      [[FAM19A5, family with sequence similarity 19 (chemokine (C-C motif)-like), member A5 [Source:HGNC Symbol;Acc:21592]]]  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df.to_csv(\"genes.csv\", index=False)\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PredClass                     Protein      \n",
       "Balancing Selection           AK1              1\n",
       "                              FAM118A          1\n",
       "                              RP11-203J24.9    1\n",
       "                              SMC1B            1\n",
       "                              TPP2             1\n",
       "Negative Freq-Dep. Selection  ADAM12           1\n",
       "                              AK1              1\n",
       "                              AL035610.1       1\n",
       "                              AL035610.2       1\n",
       "                              C6orf99          1\n",
       "                              CTD-2571L23.6    1\n",
       "                              ELFN2            1\n",
       "                              FAM118A          1\n",
       "                              FAM19A5          1\n",
       "                              IL1RAP           1\n",
       "                              PACRG            1\n",
       "                              RP1-63G5.5       1\n",
       "                              RP11-203J24.9    1\n",
       "                              RP11-648K4.2     1\n",
       "                              SLC39A11         1\n",
       "                              SMC1B            1\n",
       "                              SV2B             1\n",
       "                              TPP2             1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def collect_protein_names(df):\n",
    "    protein_data = []\n",
    "\n",
    "    for index, row in df.iterrows():\n",
    "        pred_class = row['PredClass']\n",
    "        proteins = row['Proteins']\n",
    "        \n",
    "        for protein in proteins:\n",
    "            protein_name = protein[0]\n",
    "            protein_data.append((pred_class, protein_name))\n",
    "\n",
    "    return protein_data\n",
    "\n",
    "protein_data = collect_protein_names(filtered_df)\n",
    "protein_df = pd.DataFrame(protein_data, columns=['PredClass', 'Protein'])\n",
    "result_df = protein_df.groupby(['PredClass', 'Protein']).size()\n",
    "pd.set_option('display.max_rows', None)\n",
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK1</td>\n",
       "      <td>Involved in the regulation of adenylate kinase activity and energy metabolism.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FAM118A</td>\n",
       "      <td>Potential role in cellular processes and disease mechanisms.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SMC1B</td>\n",
       "      <td>Part of the cohesin complex, important for chromosome segregation during meiosis.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SLC39A11</td>\n",
       "      <td>Zinc transporter involved in maintaining zinc homeostasis.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RP11-648K4.2</td>\n",
       "      <td>Non-coding RNA, function not well understood.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RP11-203J24.9</td>\n",
       "      <td>Non-coding RNA, potential regulatory roles.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RP1-63G5.5</td>\n",
       "      <td>Non-coding RNA, function not well understood.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PACRG</td>\n",
       "      <td>Associated with Parkinsons disease, involved in microtubule stability.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>IL1RAP</td>\n",
       "      <td>Interleukin-1 receptor accessory protein, involved in inflammatory response.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FAM19A5</td>\n",
       "      <td>Potentially involved in neural development and function.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ELFN2</td>\n",
       "      <td>Potential role in synapse formation.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CTD-2571L23.6</td>\n",
       "      <td>Non-coding RNA, function not well understood.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>C6orf99</td>\n",
       "      <td>Chromosome 6 open reading frame 99, function not well understood.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AL035610.2</td>\n",
       "      <td>Non-coding RNA, function not well understood.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AL035610.1</td>\n",
       "      <td>Non-coding RNA, function not well understood.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ADAM12</td>\n",
       "      <td>Metalloprotease involved in cell adhesion, signaling, and migration.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TPP2</td>\n",
       "      <td>Tripeptidyl peptidase II, involved in protein degradation.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SV2B</td>\n",
       "      <td>Synaptic vesicle glycoprotein involved in neurotransmitter release.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Protein  \\\n",
       "0             AK1   \n",
       "1         FAM118A   \n",
       "2           SMC1B   \n",
       "3        SLC39A11   \n",
       "4    RP11-648K4.2   \n",
       "5   RP11-203J24.9   \n",
       "6      RP1-63G5.5   \n",
       "7           PACRG   \n",
       "8          IL1RAP   \n",
       "9         FAM19A5   \n",
       "10          ELFN2   \n",
       "11  CTD-2571L23.6   \n",
       "12        C6orf99   \n",
       "13     AL035610.2   \n",
       "14     AL035610.1   \n",
       "15         ADAM12   \n",
       "16           TPP2   \n",
       "17           SV2B   \n",
       "\n",
       "                                                                             Function  \n",
       "0      Involved in the regulation of adenylate kinase activity and energy metabolism.  \n",
       "1                        Potential role in cellular processes and disease mechanisms.  \n",
       "2   Part of the cohesin complex, important for chromosome segregation during meiosis.  \n",
       "3                          Zinc transporter involved in maintaining zinc homeostasis.  \n",
       "4                                       Non-coding RNA, function not well understood.  \n",
       "5                                         Non-coding RNA, potential regulatory roles.  \n",
       "6                                       Non-coding RNA, function not well understood.  \n",
       "7             Associated with Parkinsons disease, involved in microtubule stability.  \n",
       "8        Interleukin-1 receptor accessory protein, involved in inflammatory response.  \n",
       "9                            Potentially involved in neural development and function.  \n",
       "10                                               Potential role in synapse formation.  \n",
       "11                                      Non-coding RNA, function not well understood.  \n",
       "12                  Chromosome 6 open reading frame 99, function not well understood.  \n",
       "13                                      Non-coding RNA, function not well understood.  \n",
       "14                                      Non-coding RNA, function not well understood.  \n",
       "15               Metalloprotease involved in cell adhesion, signaling, and migration.  \n",
       "16                         Tripeptidyl peptidase II, involved in protein degradation.  \n",
       "17                Synaptic vesicle glycoprotein involved in neurotransmitter release.  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteins = pd.read_csv('/Users/kristinagrigaityte/PycharmProjects/pulls/Files/BSModel/Predictions/proteins.csv')\n",
    "proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
